Remove redundant part of if statement

j-t-1 opened the pull request:
Also reorder imports.

j-t-1 committed:
Remove redundant part of if statement

Also reorder imports.

j-t-1 committed:
Remove redundant part of if statement

Also reorder imports.

j-t-1 committed:
Remove redundant part of if statement

Also reorder imports.

j-t-1 commented:
Please could you check the if statement re-write?

The original was:
```
        for keyword in kwargs:
            val = kwargs[keyword]
            if not isinstance(
                val, interfaces.configuration.BasicTypes
            ) and not isinstance(val, list):
                if not isinstance(val, list) or all(
                    isinstance(x, interfaces.configuration.BasicTypes) for x in val
                ):
                    raise TypeError(
                        "Configurable values must be simple types (int, bool, str, bytes)"
                    )
```

But maybe it should be:
 ```
        for keyword in kwargs:
            val = kwargs[keyword]
            if not isinstance(
                val, interfaces.configuration.BasicTypes
            ):
                if isinstance(val, list) and not all(
                    isinstance(x, interfaces.configuration.BasicTypes) for x in val
                ):
                    raise TypeError(
                        "Configurable values must be simple types (int, bool, str, bytes)"
                    )
```

j-t-1 commented:
1. Can this line be removed: # TODO: provide help, consider generic functions (pslist?) and/or providing windows/linux functions
2. Is the readline and rlcompleter imports needed post Python 3.4? Probably needed, but my computer is down so cannot currently check this.

gcmoreira commented on volatility3/cli/volshell/generic.py:
```diff
@@ -553,14 +553,9 @@ def create_configurable(
-                ):
+        for keyword, val in kwargs.items():
+            if not isinstance(val, (interfaces.configuration.BasicTypes, list)):
+                if all(isinstance(x, interfaces.configuration.BasicTypes) for x in val):
```
hm.. it seems the original implementation doesn't support `tuple`s, but yours does, the `if not isinstance(val, list) or` might be intentional. Additionally, if val is of type `float`, this might crash since it's neither in `BasicTypes`, a list, and is definitely not iterable. 

@ikelos On the other hand, we need to fix [this](https://github.com/volatilityfoundation/volatility3/blob/b8298a345fc5a3260ca2156d1e81a292530bde6e/volatility3/framework/interfaces/configuration.py#L788), where the values don't align with those in `BasicTypes`. Maybe, that line should use `BasicTypes` and `BasicTypes` needs to be updated too to also include `float` or ... ideally, we should have a dedicated function to centralize type validation for configuration values.

ikelos replied:
So I don't believe floats were supposed to be allowed in config items (because SimpleTypes was supposed to be single point of truth for this), so they can *probably* be removed without problem from the config type verification.  I agree we should unify that list with `BasicTypes`.  I don't think there's necessarily point in a function for verification, but a shared list definitely makes sense (and the code you pointed out should be using it, the only reason I can think why it might legitimately not be using it is if it causes import loops).

Yeah, I've no idea why the ifs seemed to be run twice, but I did write it 7 years ago.  5;P  I'm not sure the replacement works correctly.  It's basically checking that everything's a basic type, or is a list of basic types (so no lists of lists, and no other complex types involved although vol types would be allowed because they all inherit from standard base types).

ikelos replied:
The one you've got (and to be fair I think also the original) allows a list of anything to get through, which wasn't the intention).  The intention was either a `BasicType` or a list where each element is only a `BasicType` (and not a list).

j-t-1 replied:
This PR became a bit messy (the first commit allows a list of anything to get through), is it better if you close it and I do a new PR with just the change in the last commit?

From the idea of @gcmoreira above, may better to place in volatility3/framework/objects/utility.py.

j-t-1 committed:
Refactor check of BasicType

The intention is either a BasicType or a list where each element is only a BasicType (and not a list).

j-t-1 committed:
Merge branch 'volatilityfoundation:develop' into Volshell

ikelos commented on volatility3/cli/volshell/generic.py:
```diff
@@ -11,18 +11,18 @@
 except ImportError:
     has_capstone = False
 
+from volatility3.cli import text_renderer, volshell
```
I don't *think* the position of these affects anything?  The attempt to import capstone just sets a variable (which other modules shouldn't be using.  Was there a specififc reason for this shift, or just tidying up?  Generally I try to put all the code below all the imports unless there's a good reason?

j-t-1 replied:
It was to put in the order of standard libraries, 3rd party libraries, then ours. But I do not mind, and am fine with it remaining as it was. Could run the codebase through isort in the future.

ikelos commented on volatility3/cli/volshell/generic.py:
```diff
@@ -553,17 +553,16 @@ def create_configurable(
-                    )
+        for keyword, val in kwargs.items():
+            BasicType_or_list_of_BasicType = False  # excludes list of lists
+            if isinstance(val, interfaces.configuration.BasicTypes):
```
Feels like his could be compressed into an or without need for an additional variable, but I'm fine if you prefer doing it explicitly.

ikelos reviewed (commented):
Thanks, this looks ok now if you're happy with it?

j-t-1 commented:
Is it easy for you to revert the imports change, and then commit? Or let me know and can do a completely completely new PR, or an additional commit to revert the imports change. Or feel free to commit as is.
